File: benchmarks/drivers/python/bench_mongocore.py
"""Benchmark MongoCore Python client (via gRPC sidecar)."""

import asyncio
import json
import os
import platform
import sys
import time
import statistics
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "clients" / "python" / "src"))
from mongocore import MongoClient

CONFIG = json.loads((Path(__file__).parent.parent / "common.json").read_text())
DATA_DIR = Path(__file__).parent.parent.parent / "data"
RESULTS_DIR = Path(__file__).parent.parent.parent / "results"
RESULTS_DIR.mkdir(exist_ok=True)

QUICK_MODE = "--quick" in sys.argv

WARMUP = 0 if QUICK_MODE else CONFIG["warmup_iterations"]["python"]
MIN_TIME = 0 if QUICK_MODE else CONFIG["min_time_secs"]
MAX_ITERS = 1 if QUICK_MODE else CONFIG["max_iterations"]
MAX_TIME = 5 if QUICK_MODE else CONFIG["max_time_secs"]
DB_NAME = CONFIG["database"]
ADDR = CONFIG["mongocore_address"]
SOCKET_PATH = CONFIG.get("mongocore_socket_path", "/tmp/mongocore.sock")


_ACTUAL_TRANSPORT = "tcp"

def get_system_info():
    return {
        "os": platform.system().lower(),
        "arch": platform.machine(),
        "cpus": os.cpu_count(),
        "mongocore_version": "0.6.0",
        "driver": "mongocore+python",
        "transport": _ACTUAL_TRANSPORT,
    }


async def run_benchmark(name, category, setup_fn, before_task_fn, task_fn, after_task_fn, teardown_fn, dataset_size_bytes, batch_size=1):
    """Run a benchmark following MongoDB spec methodology."""
    # Force gRPC transport — this benchmark measures gRPC performance.
    # Try UDS first if socket exists, fall back to TCP on connection issues.
    client = None
    if os.path.exists(SOCKET_PATH):
        try:
            client = MongoClient(ADDR, socket_path=SOCKET_PATH, transport="grpc")
            await client.connect()
            await client.run_command("admin", {"hello": 1})
        except Exception:
            await client.close() if client else None
            client = None

    if client is None:
        client = MongoClient(ADDR, transport="grpc")
        await client.connect()
    else:
        global _ACTUAL_TRANSPORT
        _ACTUAL_TRANSPORT = "uds"

    await setup_fn(client)

    # Warmup
    for _ in range(WARMUP):
        await before_task_fn(client)
        await task_fn(client)
        await after_task_fn(client)

    # Timed iterations
    times = []
    total_time = 0.0
    iteration = 0

    while total_time < MIN_TIME or iteration < 5:
        if iteration >= MAX_ITERS or total_time >= MAX_TIME:
            break

        await before_task_fn(client)

        start = time.perf_counter()
        await task_fn(client)
        elapsed = time.perf_counter() - start

        await after_task_fn(client)

        times.append(elapsed)
        total_time += elapsed
        iteration += 1

    await teardown_fn(client)
    await client.close()

    times.sort()
    median = statistics.median(times)
    ops_per_sec = batch_size / median
    mb_per_sec = dataset_size_bytes / median / 1_000_000

    def percentile(data, pct):
        import math
        idx = max(0, math.ceil(len(data) * pct / 100) - 1)
        return data[min(idx, len(data) - 1)]

    result = {
        "benchmark": name,
        "category": category,
        "driver": "mongocore+python",
        "dataset_size_bytes": dataset_size_bytes,
        "batch_size": batch_size,
        "iterations": len(times),
        "total_time_secs": round(total_time, 3),
        "ops_per_sec": round(ops_per_sec, 1),
        "mb_per_sec": round(mb_per_sec, 3),
        "percentiles": {
            "p10": round(percentile(times, 10), 6),
            "p25": round(percentile(times, 25), 6),
            "p50": round(median, 6),
            "p75": round(percentile(times, 75), 6),
            "p90": round(percentile(times, 90), 6),
            "p95": round(percentile(times, 95), 6),
            "p99": round(percentile(times, 99), 6),
        },
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "system": get_system_info(),
    }
    print(f"  {name}: {ops_per_sec:.0f} ops/s, {mb_per_sec:.2f} MB/s ({len(times)} iters)")
    return result


async def main():
    print(f"=== MongoCore+Python benchmarks (will try UDS then TCP) ===")
    results = []

    small_doc = json.loads((DATA_DIR / "small_doc.json").read_text())
    tweet_doc = json.loads((DATA_DIR / "tweet.json").read_text())
    large_doc = json.loads((DATA_DIR / "large_doc.json").read_text())
    small_size = len(json.dumps(small_doc).encode())
    tweet_size = len(json.dumps(tweet_doc).encode())
    large_size = len(json.dumps(large_doc).encode())

    # Run Command (batch 10,000 hello commands per iteration)
    async def task_run_command(c):
        for _ in range(10_000):
            await c.run_command(DB_NAME, {"hello": 1})

    results.append(await run_benchmark(
        "run_command", "single_doc",
        setup_fn=lambda c: asyncio.sleep(0),
        before_task_fn=lambda c: asyncio.sleep(0),
        task_fn=task_run_command,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=10_000 * 100, batch_size=10_000,
    ))

    # Find One by ID (batch 10,000 finds per iteration)
    async def setup_find(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_find_mc"})
        except:
            pass
        coll = c[DB_NAME]["bench_find_mc"]
        await coll.insert_one({"_id": "bench_find_001", **tweet_doc})

    async def task_find(c):
        for _ in range(10_000):
            await c[DB_NAME]["bench_find_mc"].find_one({"_id": "bench_find_001"})

    async def teardown_find(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_find_mc"})
        except:
            pass

    results.append(await run_benchmark(
        "find_one_by_id", "single_doc",
        setup_fn=setup_find,
        before_task_fn=lambda c: asyncio.sleep(0),
        task_fn=task_find,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=teardown_find,
        dataset_size_bytes=10_000 * tweet_size, batch_size=10_000,
    ))

    # InsertOne Small (batch 10,000 inserts per iteration)
    async def task_insert_small(c):
        from bson import ObjectId
        for _ in range(10_000):
            await c[DB_NAME]["bench_insert_small_mc"].insert_one({**small_doc, "_id": str(ObjectId())})

    async def before_insert_small(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_insert_small_mc"})
        except:
            pass

    results.append(await run_benchmark(
        "insert_one_small", "single_doc",
        setup_fn=lambda c: asyncio.sleep(0),
        before_task_fn=before_insert_small,
        task_fn=task_insert_small,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=10_000 * small_size, batch_size=10_000,
    ))

    # InsertOne Large (batch 10 inserts per iteration, large docs ~2.75MB each)
    async def task_insert_large(c):
        from bson import ObjectId
        for _ in range(10):
            await c[DB_NAME]["bench_insert_large_mc"].insert_one({**large_doc, "_id": str(ObjectId())})

    async def before_insert_large(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_insert_large_mc"})
        except:
            pass

    results.append(await run_benchmark(
        "insert_one_large", "single_doc",
        setup_fn=lambda c: asyncio.sleep(0),
        before_task_fn=before_insert_large,
        task_fn=task_insert_large,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=10 * large_size, batch_size=10,
    ))

    # Bulk Insert Small (10K per iteration)
    async def task_bulk(c):
        from bson import ObjectId
        docs = [{**small_doc, "_id": str(ObjectId())} for _ in range(10_000)]
        await c[DB_NAME]["bench_bulk_mc"].insert_many(docs)

    async def before_bulk(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_bulk_mc"})
        except:
            pass

    results.append(await run_benchmark(
        "bulk_insert_small", "multi_doc",
        setup_fn=lambda c: asyncio.sleep(0),
        before_task_fn=before_bulk,
        task_fn=task_bulk,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=small_size * 10_000, batch_size=10_000,
    ))

    # Find Many (10K docs — now enabled with 64MB message limit)
    async def setup_find_many(c):
        from bson import ObjectId
        try:
            await c.run_command(DB_NAME, {"drop": "bench_find_many_mc"})
        except:
            pass
        docs = [{**small_doc, "_id": str(ObjectId())} for _ in range(10_000)]
        await c[DB_NAME]["bench_find_many_mc"].insert_many(docs)

    async def task_find_many(c):
        await c[DB_NAME]["bench_find_many_mc"].find({})

    results.append(await run_benchmark(
        "find_many", "multi_doc",
        setup_fn=setup_find_many,
        before_task_fn=lambda c: asyncio.sleep(0),
        task_fn=task_find_many,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=small_size * 10_000, batch_size=10_000,
    ))

    # Bulk Insert Large (10 x 2.75MB docs per iteration — enabled with 64MB message limit)
    async def task_bulk_large(c):
        from bson import ObjectId
        docs = [{**large_doc, "_id": str(ObjectId())} for _ in range(10)]
        await c[DB_NAME]["bench_bulk_large_mc"].insert_many(docs)

    async def before_bulk_large(c):
        try:
            await c.run_command(DB_NAME, {"drop": "bench_bulk_large_mc"})
        except:
            pass

    results.append(await run_benchmark(
        "bulk_insert_large", "multi_doc",
        setup_fn=lambda c: asyncio.sleep(0),
        before_task_fn=before_bulk_large,
        task_fn=task_bulk_large,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=large_size * 10, batch_size=10,
    ))

    # Find Many Large (10 x 2.75MB docs — enabled with 64MB message limit)
    async def setup_find_many_large(c):
        from bson import ObjectId
        try:
            await c.run_command(DB_NAME, {"drop": "bench_find_many_large_mc"})
        except:
            pass
        docs = [{**large_doc, "_id": str(ObjectId())} for _ in range(10)]
        await c[DB_NAME]["bench_find_many_large_mc"].insert_many(docs)

    async def task_find_many_large(c):
        await c[DB_NAME]["bench_find_many_large_mc"].find({})

    results.append(await run_benchmark(
        "find_many_large", "multi_doc",
        setup_fn=setup_find_many_large,
        before_task_fn=lambda c: asyncio.sleep(0),
        task_fn=task_find_many_large,
        after_task_fn=lambda c: asyncio.sleep(0),
        teardown_fn=lambda c: asyncio.sleep(0),
        dataset_size_bytes=large_size * 10, batch_size=10,
    ))

    # Save results
    output_path = RESULTS_DIR / "python_mongocore.json"
    with open(output_path, "w") as f:
        json.dump(results, f, indent=2)
    print(f"\nResults saved to {output_path}")


if __name__ == "__main__":
    asyncio.run(main())

File: benchmarks/drivers/go/bench_mongocore.go
// Benchmark MongoCore Go client (via gRPC sidecar).
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"time"

	"github.com/rozza/mongocore/clients/go/mongocore"
	"go.mongodb.org/mongo-driver/v2/bson"
	
)

type Config struct {
	MongoDBURI     string `json:"mongodb_uri"`
	MongoCoreAddr  string `json:"mongocore_address"`
	Database       string `json:"database"`
	MinTimeSecs    int    `json:"min_time_secs"`
	MaxIterations  int    `json:"max_iterations"`
	MaxTimeSecs    int    `json:"max_time_secs"`
	WarmupIters    map[string]int `json:"warmup_iterations"`
}

type SystemInfo struct {
	OS              string `json:"os"`
	Arch            string `json:"arch"`
	CPUs            int    `json:"cpus"`
	MongoCoreVer    string `json:"mongocore_version"`
	Driver          string `json:"driver"`
}

type Percentiles struct {
	P10 float64 `json:"p10"`
	P25 float64 `json:"p25"`
	P50 float64 `json:"p50"`
	P75 float64 `json:"p75"`
	P90 float64 `json:"p90"`
	P95 float64 `json:"p95"`
	P99 float64 `json:"p99"`
}

type BenchResult struct {
	Benchmark       string      `json:"benchmark"`
	Category        string      `json:"category"`
	Driver          string      `json:"driver"`
	DatasetBytes    int         `json:"dataset_size_bytes"`
	BatchSize       int         `json:"batch_size"`
	Iterations      int         `json:"iterations"`
	TotalTimeSecs   float64     `json:"total_time_secs"`
	OpsPerSec       float64     `json:"ops_per_sec"`
	MBPerSec        float64     `json:"mb_per_sec"`
	Percentiles     Percentiles `json:"percentiles"`
	Timestamp       string      `json:"timestamp"`
	System          SystemInfo  `json:"system"`
}


func getSystemInfo() SystemInfo {
	return SystemInfo{
		OS:           runtime.GOOS,
		Arch:         runtime.GOARCH,
		CPUs:         runtime.NumCPU(),
		MongoCoreVer: "0.6.0",
		Driver:       "mongocore+go",
	}
}

func percentile(data []float64, pct int) float64 {
	idx := int(math.Ceil(float64(len(data))*float64(pct)/100.0)) - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(data) {
		idx = len(data) - 1
	}
	return data[idx]
}

func runBenchmark(
	name, category string,
	setupFn func(*mongocore.Client) error,
	beforeTaskFn func(*mongocore.Client) error,
	taskFn func(*mongocore.Client) error,
	afterTaskFn func(*mongocore.Client) error,
	teardownFn func(*mongocore.Client) error,
	datasetBytes, batchSize int,
	config Config,
) BenchResult {
	client := mongocore.MongoClientTCP(config.MongoCoreAddr)
	ctx := context.Background()
	if err := client.Connect(ctx); err != nil {
		panic(err)
	}
	defer client.Close()

	if err := setupFn(client); err != nil {
		panic(err)
	}

	// Warmup
	warmup := config.WarmupIters["go"]
	for i := 0; i < warmup; i++ {
		if err := beforeTaskFn(client); err != nil {
			panic(err)
		}
		if err := taskFn(client); err != nil {
			panic(err)
		}
		if err := afterTaskFn(client); err != nil {
			panic(err)
		}
	}

	// Timed iterations
	times := []float64{}
	totalTime := 0.0
	iteration := 0

	for totalTime < float64(config.MinTimeSecs) || iteration < 5 {
		if iteration >= config.MaxIterations || totalTime >= float64(config.MaxTimeSecs) {
			break
		}

		if err := beforeTaskFn(client); err != nil {
			panic(err)
		}

		start := time.Now()
		if err := taskFn(client); err != nil {
			panic(err)
		}
		elapsed := time.Since(start).Seconds()

		if err := afterTaskFn(client); err != nil {
			panic(err)
		}

		times = append(times, elapsed)
		totalTime += elapsed
		iteration++
	}

	if err := teardownFn(client); err != nil {
		panic(err)
	}

	// Calculate metrics
	sort.Float64s(times)
	median := times[len(times)/2]
	opsPerSec := float64(batchSize) / median
	mbPerSec := float64(datasetBytes) / median / 1_000_000

	result := BenchResult{
		Benchmark:     name,
		Category:      category,
		Driver:        "mongocore+go",
		DatasetBytes:  datasetBytes,
		BatchSize:     batchSize,
		Iterations:    len(times),
		TotalTimeSecs: math.Round(totalTime*1000) / 1000,
		OpsPerSec:     math.Round(opsPerSec*10) / 10,
		MBPerSec:      math.Round(mbPerSec*1000) / 1000,
		Percentiles: Percentiles{
			P10: math.Round(percentile(times, 10)*1_000_000) / 1_000_000,
			P25: math.Round(percentile(times, 25)*1_000_000) / 1_000_000,
			P50: math.Round(median*1_000_000) / 1_000_000,
			P75: math.Round(percentile(times, 75)*1_000_000) / 1_000_000,
			P90: math.Round(percentile(times, 90)*1_000_000) / 1_000_000,
			P95: math.Round(percentile(times, 95)*1_000_000) / 1_000_000,
			P99: math.Round(percentile(times, 99)*1_000_000) / 1_000_000,
		},
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		System:    getSystemInfo(),
	}

	fmt.Printf("  %s: %.0f ops/s, %.2f MB/s (%d iterations)\n", name, opsPerSec, mbPerSec, len(times))
	return result
}

func isQuickMode() bool {
	for _, arg := range os.Args[1:] {
		if arg == "--quick" {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("=== MongoCore+Go benchmarks ===")

	// Load config
	configPath := filepath.Join("..", "common.json")
	configBytes, err := os.ReadFile(configPath)
	if err != nil {
		panic(err)
	}
	var config Config
	if err := json.Unmarshal(configBytes, &config); err != nil {
		panic(err)
	}

	if isQuickMode() {
		config.WarmupIters["go"] = 0
		config.MinTimeSecs = 0
		config.MaxIterations = 1
		config.MaxTimeSecs = 5
	}

	// Load test documents
	dataDir := filepath.Join("..", "..", "data")
	smallDocBytes, err := os.ReadFile(filepath.Join(dataDir, "small_doc.json"))
	if err != nil {
		panic(fmt.Sprintf("Failed to read small_doc.json: %v", err))
	}
	tweetDocBytes, err := os.ReadFile(filepath.Join(dataDir, "tweet.json"))
	if err != nil {
		panic(fmt.Sprintf("Failed to read tweet.json: %v", err))
	}
	largeDocBytes, err := os.ReadFile(filepath.Join(dataDir, "large_doc.json"))
	if err != nil {
		panic(fmt.Sprintf("Failed to read large_doc.json: %v", err))
	}

	var smallDoc, tweetDoc, largeDoc map[string]interface{}
	json.Unmarshal(smallDocBytes, &smallDoc)
	json.Unmarshal(tweetDocBytes, &tweetDoc)
	json.Unmarshal(largeDocBytes, &largeDoc)

	smallSize := len(smallDocBytes)
	tweetSize := len(tweetDocBytes)
	largeSize := len(largeDocBytes)

	ctx := context.Background()
	results := []BenchResult{}

	// Run Command (batch 10,000 hello commands per iteration)
	results = append(results, runBenchmark(
		"run_command", "single_doc",
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			for i := 0; i < 10_000; i++ {
				if _, err := c.RunCommand(ctx, config.Database, bson.D{{Key: "hello", Value: 1}}, false); err != nil {
					return err
				}
			}
			return nil
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		10_000*100, 10_000, config,
	))

	// Find One by ID (batch 10,000 finds per iteration)
	results = append(results, runBenchmark(
		"find_one_by_id", "single_doc",
		func(c *mongocore.Client) error {
			doc := bson.D{{Key: "_id", Value: "bench_find_001"}}
			for k, v := range tweetDoc {
				if k != "_id" {
					doc = append(doc, bson.E{Key: k, Value: v})
				}
			}
			// Drop and recreate
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_find_mc"}}, false)
			_, err := c.Database(config.Database).Collection("bench_find_mc").InsertOne(ctx, doc)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			for i := 0; i < 10_000; i++ {
				if _, err := c.Database(config.Database).Collection("bench_find_mc").FindOne(ctx, bson.D{{Key: "_id", Value: "bench_find_001"}}); err != nil {
					return err
				}
			}
			return nil
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_find_mc"}}, false)
			return nil
		},
		10_000*tweetSize, 10_000, config,
	))

	// InsertOne Small (batch 10,000 inserts per iteration)
	results = append(results, runBenchmark(
		"insert_one_small", "single_doc",
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_insert_small_mc"}}, false)
			return nil
		},
		func(c *mongocore.Client) error {
			for i := 0; i < 10_000; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range smallDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				if _, err := c.Database(config.Database).Collection("bench_insert_small_mc").InsertOne(ctx, doc); err != nil {
					return err
				}
			}
			return nil
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		10_000*smallSize, 10_000, config,
	))

	// InsertOne Large (batch 10 inserts per iteration, large docs ~2.75MB each)
	results = append(results, runBenchmark(
		"insert_one_large", "single_doc",
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_insert_large_mc"}}, false)
			return nil
		},
		func(c *mongocore.Client) error {
			for i := 0; i < 10; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range largeDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				if _, err := c.Database(config.Database).Collection("bench_insert_large_mc").InsertOne(ctx, doc); err != nil {
					return err
				}
			}
			return nil
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		10*largeSize, 10, config,
	))

	// Bulk Insert Small (10K per iteration)
	results = append(results, runBenchmark(
		"bulk_insert_small", "multi_doc",
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_bulk_mc"}}, false)
			return nil
		},
		func(c *mongocore.Client) error {
			docs := make([]bson.D, 10_000)
			for i := 0; i < 10_000; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range smallDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				docs[i] = doc
			}
			_, err := c.Database(config.Database).Collection("bench_bulk_mc").InsertMany(ctx, docs)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		smallSize*10_000, 10_000, config,
	))

	// Find Many (10K docs — enabled with 64MB message limit)
	results = append(results, runBenchmark(
		"find_many", "multi_doc",
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_find_many_mc"}}, false)
			docs := make([]bson.D, 10_000)
			for i := 0; i < 10_000; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range smallDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				docs[i] = doc
			}
			_, err := c.Database(config.Database).Collection("bench_find_many_mc").InsertMany(ctx, docs)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			_, err := c.Database(config.Database).Collection("bench_find_many_mc").Find(ctx, bson.D{}, nil)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		smallSize*10_000, 10_000, config,
	))

	// Bulk Insert Large (10 x 2.75MB docs — enabled with 64MB message limit)
	results = append(results, runBenchmark(
		"bulk_insert_large", "multi_doc",
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_bulk_large_mc"}}, false)
			return nil
		},
		func(c *mongocore.Client) error {
			docs := make([]bson.D, 10)
			for i := 0; i < 10; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range largeDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				docs[i] = doc
			}
			_, err := c.Database(config.Database).Collection("bench_bulk_large_mc").InsertMany(ctx, docs)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		largeSize*10, 10, config,
	))

	// Find Many Large (10 x 2.75MB docs — enabled with 64MB message limit)
	results = append(results, runBenchmark(
		"find_many_large", "multi_doc",
		func(c *mongocore.Client) error {
			c.RunCommand(ctx, config.Database, bson.D{{Key: "drop", Value: "bench_find_many_large_mc"}}, false)
			docs := make([]bson.D, 10)
			for i := 0; i < 10; i++ {
				doc := bson.D{{Key: "_id", Value: bson.NewObjectID().Hex()}}
				for k, v := range largeDoc {
					if k != "_id" {
						doc = append(doc, bson.E{Key: k, Value: v})
					}
				}
				docs[i] = doc
			}
			_, err := c.Database(config.Database).Collection("bench_find_many_large_mc").InsertMany(ctx, docs)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error {
			_, err := c.Database(config.Database).Collection("bench_find_many_large_mc").Find(ctx, bson.D{}, nil)
			return err
		},
		func(c *mongocore.Client) error { return nil },
		func(c *mongocore.Client) error { return nil },
		largeSize*10, 10, config,
	))

	// Save results
	resultsDir := filepath.Join("..", "..", "results")
	os.MkdirAll(resultsDir, 0755)
	outputPath := filepath.Join(resultsDir, "go_mongocore.json")
	resultsJSON, _ := json.MarshalIndent(results, "", "  ")
	if err := os.WriteFile(outputPath, resultsJSON, 0644); err != nil {
		panic(err)
	}
	fmt.Printf("\nResults saved to %s\n", outputPath)
}
